addons/stock_account/models/account_move.py

```diff
@@ -45,8 +45,14 @@ def _post(self, soft=True):
         if self._context.get('move_reverse_cancel'):
             return super()._post(soft)
 
+        to_post = self
+        if soft:
+            # If soft is True, then a forward dated invoice will NOT be posted now.
+            # Rather, it will set "auto_post" to True - it will get anglo saxon lines at that post date.
+            to_post -= self.filtered(lambda move: move.date > fields.Date.context_today(self))
+
         # Create additional COGS lines for customer invoices.
-        self.env['account.move.line'].create(self._stock_account_prepare_anglo_saxon_out_lines_vals())
+        self.env['account.move.line'].create(to_post._stock_account_prepare_anglo_saxon_out_lines_vals())
```

Comment:
wouldn't be easier to filter out account move of `self` with auto_post=True into `_stock_account_prepare_anglo_saxon_out_lines_vals` ?

Reply:
@Whenrow 

But auto_post is not set yet.  That method has no knowledge of whether it will post or not?  It's role is to create moves when asked.

THIS method is to ensure that the method is called when a move is posted.  Unfortunately, the problem is that not every record in the recordset WILL post.

Some other override methods to the correct thing, by calling super first - but this cannot be done in this case because the lines are needed before super is called.

E.g

```
    def _post(self, soft=True):
        # OVERRIDE
        # On the validation of an invoice, send the different corrected fields to iap to improve the ocr algorithm.
        posted = super()._post(soft)
        ....

```

### The ideal

TBH, the correct solution is that _post in account_move should be a wrapper method, which pre-parses those which wont post, and then calls a different method to effect the actual posting, but that is quite a significant restructure.

Most of the following have the same problem, because they all make data changes before calling the super, on the assumption that the records will ALL post.

  - odoo/addons/purchase_stock/models/account_invoice.py, this is the reverse of this fix
  - enterprise/account_taxcloud/models/account_invoice.py, performs api calls on the invoices.
  - enterprise/account_reports/models/account_activity.py, it closes tax entries, and generates new move line records, for transactions (not just invoices) which might not post.
  - enterprise/l10n_ar_edi/models/account_move.py, similar problems, possibly, although it seems to all _post on each record it is dealing with.
  - enterprise/l10n_it_reports/models/account_move.py, definitely suspect.
  - enterprise/l10n_mx_edi/models/account_move.py, might be a problem, might be harmless.
  - odoo/addons/l10n_id_efaktur/models/account_move.py like the one above, may be harmless or not.
  - odoo/addons/l10n_sa_invoice/models/account_move.py is kind of correct, but then processes all moves, not just the ones which were posted.

### Additionally:

Many methods that override _post show errors for various non-posting reasons.  ALL of these are just plain wrong if _post calls a transaction with an error from the cron process!

Reply:
@richard-willdooit I'm 100% okay with the fact the `_post` method should be called on account move that will be effectively posted. The filter on date should be done elsewhere. (cc @odoo/rd-accounting)

My solution consider the default value of `auto_post` to `False` to create the anglo saxon account move lines only the first time `_post()` is called, not the second via the cron.

Reply:
@Whenrow 

While that appears cleaner, this will not solve, as it means a few things.

1 - if the auto_post is consciously set by a user, then this routine will only see the invoice once, when it is actually posted by the cron, and will never get anglo saxon lines.
2 - it will get the wrong cost on the anglo saxon lines, not getting the value when the invoice is actually posted.
3 - it will mean that an unposted invoice will have angl saxon lines - so if a user visits the invoice, and unticks the auto post, it will not delete the anglo saxon lines, - they could then change the date and post it - and the same problem would happen again.

The anglo saxon lines really should not be generated until the actual post will happen.

Reply:
@william-andre 

Yes, while that is an important question, (and I do agree with your answer too), creating the extra lines at the first time leaves the transaction in an odd state in relation to others, too.

A refactor of _post is what I believe is needed, but was not in a position to propose from outside of Odoo.  As you say, that would allow for rectification and standardisation of other calls to _post.

Reply:
@william-andre @Whenrow 

Any chance this will be accepted?

Reply:
Another solution to avoid a refactoring in `account` could be to extend `write`.
```python
def write(self, vals):
    if vals.get('state') == 'posted':
        to_post = self.filtered(lambda m: not m.posted_before)
        self.env['account.move.line'].create(to_post._stock_account_prepare_anglo_saxon_out_lines_vals())
    return super().write(vals)
```

Note that it might also require a test